[k8sprocessor] Add ability to associate metadata tags using pod UID rather than just IP

pmatyjasek-sumo opened the pull request:
**Description:**
Add ability to associate metadata tags in k8s processor using pod UID rather than just IP

**Link to tracking Issue:** #1146 

**Testing:** 
Unit tests

**Documentation:** 
In-code comments

pmatyjasek-sumo committed:
Add ability to associate metadata tags in k8s processor using pod UID rather than just IP

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

jpkrohling commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	// If pod association is not set
+	if len(associations.Associations) == 0 {
```
Most of this branch seems to not have a unit test coverage. Can you improve that?

jpkrohling commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+			}
+			if asso.Name == "ip" {
+				switch asso.From {
+				case "label":
```
```suggestion
				case "labels":
```

Perhaps use a custom type for the values?

jpkrohling commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,27 @@
-// that sent the telemetry data.
-// If a match is found, the cached metadata is added to the data as resource attributes.
+// extracted metadata to the relevant spans, metrics and logs. The processor uses the kubernetes API to discover all pods
+// running in a cluster, keeps a record of their IP addresses, pod UIDs and interesting metadata.
```
> and interesting metadata

I think I've seen only pod UID and IP addresses here. Am I missing something?

pmatyjasek-sumo replied:
UID and IP are used to associate metadata with spans metrics and logs. But this metadata contains other information like Pod Name, Deployment etc.

jpkrohling replied:
Perhaps it's not part of this PR, but how would I use the pod name / deployment name and so on?

jpkrohling reviewed (commented):
Looks really nice!

pmm-sumo commented on processor/k8sprocessor/kube/client.go:
```diff
@@ -279,21 +275,48 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
+					return
+				}
+			}
+			c.Pods[pod.Status.PodIP] = newPod
```
I actually discussed this with @pmatyjasek-sumo while he was working on the extension :) Having a single map and two references to the pod just makes the things easier, since this are different domains that have no risk of conflicts

pmatyjasek-sumo replied:
@jpkrohling as @pmm-sumo mentioned we had some discussion on that and we decided that it should be ok if 2 keys from map will point to the same reference. With this we avoid to use two loops to iterate through two lists of Pods (one by IP and second by UID)

jpkrohling replied:
Sure, that's fine, but you might want to add the current size of this map in a metric somewhere: this is the kind of data that can easily explode in terms of number of elements. It's not unusual to have tens of thousands of pods in a given cluster, so, having this info would help people running the processor to identify any possible memory issue.

jpkrohling replied:
Was this done? If you don't to do it in this PR, create an issue to track this.

pmatyjasek-sumo committed:
Apply suggestions from code review

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

pmatyjasek-sumo committed:
Changes after review:
- Add more tests for added code
- Refactor Association table
- Remove unnecessary annonymous function

jpkrohling commented:
What's the current state of this PR? Is this ready to be reviews again?

pmatyjasek-sumo commented:
@jpkrohling yes it is ready for review. I've addressed previous comments in latest commit.

jpkrohling commented:
I have only one concern about the size the map might get, as it grows twice the rate of the number of pods being used. I believe the size of this map should be part of the set of metrics for the collector, as it would help correlate a possible increase in memory usage with the number of items in this map.

pmatyjasek-sumo commented:
@jpkrohling that is a good idea, I'll put this metric soon

pmatyjasek-sumo committed:
Add metric to monitor pods map size.
otelsvc/k8s/pod_table_size

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

pmatyjasek-sumo commented:
I've updated PR

jpkrohling commented on processor/k8sprocessor/observability/observability.go:
```diff
@@ -30,13 +30,15 @@ func init() {
+	mPodsUpdated  = stats.Int64("otelsvc/k8s/pod_updated", "Number of pod update events received", "1")
+	mPodsAdded    = stats.Int64("otelsvc/k8s/pod_added", "Number of pod add events received", "1")
+	mPodsDeleted  = stats.Int64("otelsvc/k8s/pod_deleted", "Number of pod delete events received", "1")
+	mPodTableSize = stats.Int64("otelsvc/k8s/pod_table_size", "Size of table containing pod info", "1")
```
Looks good, but is there a list of those metrics somewhere, telling users of this processor what they actually mean? Also, it would be a good opportunity to:

- s/otelsvc/otelcol
- replace the slashes by underscores, as it seems to be the standard at other metrics

If you don't want to do them here, feel free to open a tracking issue.

pmatyjasek-sumo replied:
@jpkrohling I've created an issue and also I'll start work on it https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2322

jpkrohling reviewed (commented):
LGTM, just a minor comment about the metrics. If you don't want to work on them here, let me know and I'll approve the PR once the tracking issue is created.

pmm-sumo commented:
@tigrannajaryan this is the PR we discussed during Log SIG yesterday. It will allow us to tag logs with K8s metadata

owais commented on processor/k8sprocessor/processor.go:
```diff
@@ -92,56 +95,31 @@ func (kp *kubernetesprocessor) ProcessMetrics(ctx context.Context, md pdata.Metr
-
-	// If IP is still not available by this point, nothing can be tagged here. Return.
-	if podIP == "" {
+	podAttributes := k8sPodAssociationFromAttributes(ctx, resource.Attributes(), kp.podAssociations)
```
We have another `attrsToAdd := kp.getAttributesForPod()` below and this one is named `podAttributes` as well. This could be a bit confusing to follow. IIRC, `podAttributes` here is actually a map of pod identifiers, such as IP, UID, etc, right? If so, let's name it `podIdentifiers` or something similar.

owais reviewed (commented):
LGTM with a couple of minor comments.

My understanding is that if a user does not update config to specify the new pod association rules, the processor will continue to work with connection/attribute IP, right?

pmatyjasek-sumo commented:
> LGTM with a couple of minor comments.
> 
> My understanding is that if a user does not update config to specify the new pod association rules, the processor will continue to work with connection/attribute IP, right?

Yes it will work as previous

pmatyjasek-sumo commented on processor/k8sprocessor/kube/client.go:
```diff
@@ -279,21 +283,46 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
+			id:   string(pod.UID),
 			name: pod.Name,
 			ts:   time.Now(),
 		})
```
good catch

tigrannajaryan commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,27 @@
-// If a match is found, the cached metadata is added to the data as resource attributes.
+// extracted metadata to the relevant spans, metrics and logs. The processor uses the kubernetes API to discover all pods
+// running in a cluster, keeps a record of their IP addresses, pod UIDs and interesting metadata.
+// The rules for associating the source record with specific Pod Metadata are configured via "pod_association" key.
```
"source record" is not an entirely clear term. Does this refer to the telemetry in pdata that is passing through the processor?

tigrannajaryan commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,27 @@
+// It represents a list of rules that are executed in order until the first one is able to do the match.
+// Each rule is specified as a pair of from (representing the rule type) and name (representing the extracted key name).
+// Following rule types are available:
+//   from: "labels" - takes the attribute from resource attributes (the value can contain either IP address or Pod UID)
```
Does "resource" here refer to pdata.Resource or to the pod? Is this pdata.Resource attributes or pod labels? I am confused by mixed usage of "labels" and "attributes" here.

pmatyjasek-sumo replied:
Ok I'll clarify that

tigrannajaryan commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,27 @@
+// Each rule is specified as a pair of from (representing the rule type) and name (representing the extracted key name).
+// Following rule types are available:
+//   from: "labels" - takes the attribute from resource attributes (the value can contain either IP address or Pod UID)
+//   from: "connection" - takes the IP attribute from connection context (if available) and automatically
```
is "ip" the only available "name" for "connection" type?

pmatyjasek-sumo replied:
Yes it is. This implementation is based on this proposal https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1146#issuecomment-702666502

tigrannajaryan commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,27 @@
+//   from: "connection" - takes the IP attribute from connection context (if available) and automatically
+//     associates it with "k8s.pod.ip" attribute
+// Pod association configuration.
+// pod_association:
```
What happens if "pod_association" setting is missing or is present but is empty?

owais replied:
I think in that case it should continue to work as it does today i.e fallback to incoming connection IP and IP attribute on resource but will let @pmatyjasek-sumo confirm.

tigrannajaryan replied:
Can we document that behavior?

pmatyjasek-sumo replied:
Ok I'll update it

pmatyjasek-sumo committed:
Address review comments:
- Provide additional description to pod association,
- Rename missleading variable
- Refactor duplicated code

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

mx-psi commented on processor/k8sprocessor/processor.go:
```diff
@@ -27,8 +26,11 @@ import (
+	k8sPodUIDLabelName string = "k8s.pod.uid"
+	podUIDLabelName    string = "pod_uid"
+	clientIPLabelName  string = "ip"
+	hostnameLabelName  string = "host.hostname"
```
open-telemetry/opentelemetry-specification#838 removed `host.hostname` from the spec in favor of `host.name`. I would suggest using `host.name`.

On a related note, some of these are available on the `go.opentelemetry.io/collector/translator/conventions` package as constant, it's better to use those if possible.

pmatyjasek-sumo committed:
Use conventions instead of constant names

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

pmatyjasek-sumo commented:
@tigrannajaryan @owais @mx-psi review has been addressed. UT failde but on `statsdreceiver` part which is not changed in this PR

mx-psi commented on processor/k8sprocessor/config_test.go:
```diff
@@ -89,5 +89,27 @@ func TestLoadConfig(t *testing.T) {
+				},
+				{
+					From: "labels",
+					Name: "host.hostname",
```
```suggestion
					Name: "host.name",
```

mx-psi commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,30 @@
+//  - from: labels
+//    name: k8s.pod.ip
+//  - from: labels
+//    name: host.hostname
```
```suggestion
//    name: host.name
```

mx-psi commented on processor/k8sprocessor/testdata/config.yaml:
```diff
@@ -47,6 +47,18 @@ processors:
+      - from: labels
+        name: k8s.pod.ip
+      - from: labels
+        name: host.hostname
```
```suggestion
        name: host.name
```

mx-psi reviewed (changes requested):
We need to change `host.hostname` in a few other places

pmatyjasek-sumo committed:
Apply suggestions from code review

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

mx-psi reviewed (approved):
LGTM!

tigrannajaryan commented on processor/k8sprocessor/processor_test.go:
```diff
@@ -286,10 +286,24 @@ func withHostname(hostname string) generateResourceFunc {
 
+func withPodUID(uid string) generateResourceFunc {
+	return func(res pdata.Resource) {
+		res.Attributes().InsertString("pod_uid", uid)
```
Is "pod_uid" some well known value?

pmatyjasek-sumo replied:
It is just example

tigrannajaryan commented on processor/k8sprocessor/processor.go:
```diff
@@ -28,6 +27,7 @@ import (
 
 const (
 	k8sIPLabelName    string = "k8s.pod.ip"
+	podUIDLabelName   string = "pod_uid"
```
What is this? Is this some magic value?

tigrannajaryan commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,30 @@
+// Pod association configuration.
+// pod_association:
+//  - from: labels
+//    name: ip
```
Do we have anywhere documented the possible values for "name"? Is it a limited list?

pmatyjasek-sumo replied:
Now it is possible to add any name that will match the label.

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+
+	}
+
+	for _, asso := range associations {
```
I have a hard time understanding what this loop does with all those ifs and map lookups. The logic escapes me.

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+				podAssociation[k8sIPLabelName] = podIP
+				continue
+			}
+			if asso.Name == "ip" {
```
I am confused. Why do we support both "ip" and "k8s.pod.ip" as the "Name" of the association? Is there any difference and which one should be used by the user if there is a difference?

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+	if len(associations) == 0 {
+		if podIP != "" {
+			podAssociation[k8sIPLabelName] = podIP
+		} else if clientIP != "" {
+			podAssociation[k8sIPLabelName] = clientIP
+		} else if contextIP != "" {
+			podAssociation[k8sIPLabelName] = contextIP
+		}
```
From what I understand contextIP is the peer ip based on connection. What is clientIP?

owais commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+)
+
+// k8sPodAssociationFromAttributes extracts IP and pod UID from attributes
+func k8sPodAssociationFromAttributes(ctx context.Context, attrs pdata.AttributeMap, associations []kube.Association) map[string]string {
```
I wonder if it'd make things cleaner if we removed the special case where association rules are not defined and instead we had a default association config that'd mimic the special case. This function would probably read better with much less conditions. WDYT?

pmatyjasek-sumo replied:
My idea was to have all stuff related to extracting identifiers in one function.

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+		if asso.Name == podUIDLabelName {
+			uid := stringAttributeFromMap(attrs, asso.Name)
+			if uid != "" {
+				podAssociation[conventions.AttributeK8sPodUID] = uid
```
This does not check if `podAssociation[conventions.AttributeK8sPodUID]` is already populated. I understand there is only one way to populate it today, but what happens if tomorrow we add another (like we have multiple ways to populate ip today).

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,98 @@
+			}
+
+		}
+	}
```
This loop and logic in it are quite difficult to understand. Please rewrite to make it clear what is happening here.

tigrannajaryan commented on processor/k8sprocessor/processor.go:
```diff
@@ -92,56 +93,31 @@ func (kp *kubernetesprocessor) ProcessMetrics(ctx context.Context, md pdata.Metr
+				resource.Attributes().InsertString(key, val)
+			}
+		}
+		resource.Attributes().InsertString(k, v)
```
Previously in passthroughMode we were only adding pod ip. Now it seems we are adding everything we know about the pod. Why the change in logic? Is this intentional?

tigrannajaryan replied:
@pmatyjasek-sumo Can you clarify this please?

pmatyjasek-sumo replied:
It behaves the same as previous. Pod attributes are added only when `passthtoughMode` is off. Line 114 adds only identifier to batch. Same as previously.

tigrannajaryan reviewed (changes requested):
Sorry, blocking since there are parts in the code that are difficult for me to understand.

tigrannajaryan commented on processor/k8sprocessor/config.go:
```diff
@@ -39,6 +39,8 @@ type Config struct {
 	// pods by labels, fields, namespaces, nodes, etc.
 	Filter FilterConfig `mapstructure:"filter"`
+
+	Association []PodAssociationConfig `mapstructure:"pod_association"`
```
Add a comment to explain what this config option does.

tigrannajaryan commented on processor/k8sprocessor/config.go:
```diff
@@ -173,3 +175,15 @@ type FieldFilterConfig struct {
 	Op string `mapstructure:"op"`
 }
+
+// PodAssociationConfig allows specifying rules for associating resources
```
Why are we mentioning "resources" here? Does "resource" here have the meaning of the "Resource" as OpenTelemetry's understands it?

tigrannajaryan commented on processor/k8sprocessor/config.go:
```diff
@@ -173,3 +175,15 @@ type FieldFilterConfig struct {
+	From string `mapstructure:"from"`
+
+	// Name represents the name of the association.
+	// e.g. ip, pod_uid
```
What are the possible values? Is "ip" and "pod_uid" the only possible values?

pmatyjasek-sumo replied:
No, there are few more

tigrannajaryan commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,30 @@
+// running in a cluster, keeps a record of their IP addresses, pod UIDs and interesting metadata.
+// The rules for associating the source record with specific Pod Metadata are configured via "pod_association" key.
+// It represents a list of rules that are executed in order until the first one is able to do the match.
+// Each rule is specified as a pair of from (representing the rule type) and name (representing the extracted key name).
```
Is the "name" the "extracted key name" or "association name" as it is written in config.go? Mixed vocabulary confuses me.

tigrannajaryan commented on processor/k8sprocessor/kube/client.go:
```diff
@@ -109,6 +117,8 @@ func (c *WatchClient) handlePodAdd(obj interface{}) {
 		c.logger.Error("object received was not of type api_v1.Pod", zap.Any("received", obj))
 	}
+	podTableSize := len(c.Pods)
+	observability.RecordPodTableSize(int64(podTableSize))
```
👍  for improving observability.

tigrannajaryan commented on processor/k8sprocessor/kube/client.go:
```diff
@@ -45,18 +45,26 @@ type WatchClient struct {
-	Pods    map[string]*Pod
-	Rules   ExtractionRules
-	Filters Filters
+	Pods         map[string]*Pod
```
Add a comment to explain what the key of the map is. I think it has a different meaning now, right?

pmatyjasek-sumo replied:
Previously the key was IP address, currently it can be either IP and UID. I'll put comment here

tigrannajaryan commented on processor/k8sprocessor/kube/kube.go:
```diff
@@ -70,7 +71,7 @@ type Pod struct {
 
 type deleteRequest struct {
-	ip   string
+	id   string
 	name string
```
Add comments to explain these 2 fields.

tigrannajaryan commented on processor/k8sprocessor/kube/kube.go:
```diff
@@ -70,7 +71,7 @@ type Pod struct {
 type deleteRequest struct {
-	ip   string
+	id   string
 	name string
```
Is this pod name? If yes can we name it `podName`?

pmatyjasek-sumo replied:
No, this is identifier (ip or uid)

tigrannajaryan commented:
@pmatyjasek-sumo thank you for working on this PR. It is extremely needed addition and highly welcome change.
I had a bit of a hard time reviewing the changes since I am poor at reading code. Please help me by simplifying the code and making it easier to understand.

pmatyjasek-sumo commented:
@tigrannajaryan Thanks for such precise review. I'll address your comments :)

pmatyjasek-sumo commented on processor/k8sprocessor/config.go:
```diff
@@ -173,3 +175,15 @@ type FieldFilterConfig struct {
+	// Allowed values are connection and labels
+	From string `mapstructure:"from"`
+
+	// Name represents the name of the association.
```
It is name of association. For example `pod_uid`, `k8s.pod.ip` etc.

tigrannajaryan replied:
In the answer below you wrote:

>name (representing the extracted key name).

If "name" is extracted key name (which I think means the attribute name if I understand correctly) then what is "name of association"?

pmatyjasek-sumo replied:
Oh, my bad. You're right it is extracted key name.

pmatyjasek-sumo committed:
Address review comments.
* Add more descriptive in-code comments for PodAssociation,
* Simplify PodAssociation logic
* Update tests

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

pmatyjasek-sumo commented:
@tigrannajaryan I've addressed Your comments. Please check if it is clean now.

tigrannajaryan commented on processor/k8sprocessor/config.go:
```diff
@@ -173,3 +177,15 @@ type FieldFilterConfig struct {
+// with logs, spans and metrics
+type PodAssociationConfig struct {
+	// From represents the source of the association.
+	// Allowed values are connection and labels
```
```suggestion
	// Allowed values are "connection" and "labels".
```

tigrannajaryan commented on processor/k8sprocessor/processor.go:
```diff
@@ -92,56 +92,31 @@ func (kp *kubernetesprocessor) ProcessMetrics(ctx context.Context, md pdata.Metr
-	attrs := resource.Attributes()
-	for k, v := range attrsToAdd {
-		attrs.InsertString(k, v)
+	for k, v := range podIdentifiers {
```
If I understand this loop correctly it may find the same pod more than once and add try to insert `attrsToAdd` more than once (if len(podIdentifiers)>1) which is unnecessary work since the second and subsequent insertion attempts are noop.

Worse, it may find different pods if `podIdentifiers` happens to identify different pods for whatever reason and in that case it will insert attributes of different pods into this `resource`.

Shouldn't this process stop once the first match happens and the pod is found (as it is described in the docs)? The problem is that breaking this loop is not going to achieve that since the order of iteration here is different from the order of rules in `kp.podAssociations`.

Generally, I don't understand why instead of having `k8sPodAssociationFromAttributes` find all identifiers and then iterate over all identifiers here, we don't iterate over `kp.podAssociations` and break after the first match. That I think would result in the behavior that matches the docs.

tigrannajaryan reviewed (commented):
Thank you @pmatyjasek-sumo 
I think I understand the code a bit bitter, but would like a bit more clarifications if possible.

pmatyjasek-sumo committed:
Apply suggestions from code review

Co-authored-by: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>

pmatyjasek-sumo commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,30 @@
+//  - from: connection
+//    name: ip
+//  - from: labels
+//    name: pod_uid
```
It is just example

tigrannajaryan replied:
I may be wrong but if I understand correctly we cannot expect to see "pod_uid" as a Resource attribute in normal use cases. Let's use an attribute name that is realistic, even if it is just an example.

pmatyjasek-sumo committed:
Code refactor:
* add custom type PodIdentifier
* fix tags adding behavior
* update comments
* refactor pod association extracting

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

pmatyjasek-sumo commented:
@tigrannajaryan I've provided some refactor to IDs extraction and association metadata tags. I hope it is clean now

tigrannajaryan commented on processor/k8sprocessor/doc.go:
```diff
@@ -15,13 +15,30 @@
+//     associates it with "k8s.pod.ip" attribute
+// Pod association configuration.
+// pod_association:
+//  - from: labels
```
I see above we now have `from: "resource_attribute"` but elsewhere we are stilling using "labels". Can we update "labels" to "resource_attribute" everywhere?

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,84 @@
+
+// extractPodIds extracts IP and pod UID from attributes or request context.
+// It returns a value pair containing configured label and IP Address and/or Pod UID
+func extractPodIds(ctx context.Context, attrs pdata.AttributeMap, associations []kube.Association) (string, kube.PodIdentifier) {
```
Please use named return values otherwise it is not clear what is the `string` that is the first returned value.

Since this returns one ID, I think it is better to use singular function name, i.e. `extractPodId`.

tigrannajaryan commented on processor/k8sprocessor/processor.go:
```diff
@@ -92,56 +92,31 @@ func (kp *kubernetesprocessor) ProcessMetrics(ctx context.Context, md pdata.Metr
-
-	// If IP is still not available by this point, nothing can be tagged here. Return.
-	if podIP == "" {
+	podIdentifierLabel, podIdentifierValue := extractPodIds(ctx, resource.Attributes(), kp.podAssociations)
```
I think it is best not to use the term "label" in this context. OpenTelemetry's vocabulary for Resources is "attribute name" or "attribute key". I think the following would be clear:
```suggestion
	podIdentifierKey, podIdentifierValue := extractPodIds(ctx, resource.Attributes(), kp.podAssociations)
```

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,84 @@
+		} else if net.ParseIP(hostname) != nil {
+			return k8sIPLabelName, kube.PodIdentifier(hostname)
+		}
+		return "", kube.PodIdentifier("")
```
Please document what it means to return empty strings.

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,84 @@
+		if asso.From == "connection" && connectionIP != "" {
+			return k8sIPLabelName, connectionIP
+		} else if asso.From == "labels" { // If association configured by labels
+			// In k8s environment, host.name label set to a pod IP address.
```
>In k8s environment, host.name label set to a pod IP address.

Where is this done? Who is setting the value of "host.name" to the pod IP address? Is it in our code? I want to understand where this assumption is coming from.

pmatyjasek-sumo replied:
I've based on previous implementation, and there was such mechanism with the comment. 
```
// Most of the metric receivers uses "host.hostname" resource label (which is represented as
		// Node.Identifier.HostName in OpenCensus format) to identify metrics origin.
		// In k8s environment, it's set to a pod IP address. If the value doesn't represent
		// an IP address, we skip it.
		if podIP == "" && md.Node.GetIdentifier().GetHostName() != "" {
```
It was introduced by this PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/358
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/358/files#diff-d980cc2893cf804127a14a1a6d9a552a69a43bc76400dffdd155bb385b31c040R197-R200

tigrannajaryan replied:
OK, thanks.

tigrannajaryan commented on processor/k8sprocessor/pod_association.go:
```diff
@@ -0,0 +1,84 @@
+					return k8sIPLabelName, kube.PodIdentifier(hostname)
+				}
+			} else {
+				// Extract values based od configured labels.
```
```suggestion
				// Extract values based on configured labels.
```

tigrannajaryan commented:
> @tigrannajaryan I've provided some refactor to IDs extraction and association metadata tags. I hope it is clean now

@pmatyjasek-sumo yes, it looks much better, thanks. Please address the remaining comments.

pmatyjasek-sumo committed:
Improve readability of the pod association code:
* Additional comments,
* Rename methods,
* fix misspells

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

pmatyjasek-sumo commented:
@tigrannajaryan please take a look

tigrannajaryan reviewed (approved):
LGTM.
Thank you @pmatyjasek-sumo
